lms/models/lti_role.py

```diff
@@ -15,14 +20,53 @@ class RoleType(str, Enum):
     ADMIN = "admin"
 
     @classmethod
-    def parse_lti_role(cls, role):
+    def parse_lti_role_type(cls, role):
         """Parse an LTI role string into one of our role types."""
 
         # We have to do this work around because Enums can't have private
         # attributes. They are just interpreted as extra values for the enum.
         return _RoleParser.parse_role(role)
 
 
+@unique
+class RoleScope(str, Enum):
+    """Enum for the different types of role a user can have."""
+
+    COURSE = "course"
+    INSTITUTION = "institution"
+    SYSTEM = "system"
+
+    @classmethod
+    def parse_lti_role_scope(cls, role):
+        """Parse an LTI role string into one of our role scopes."""
+
+        scopes_prefixes = {
+            # LTI 1.1 https://www.imsglobal.org/specs/ltiv1p0/implementation-guide
+            "urn:lti:sysrole:ims": cls.SYSTEM,
+            "urn:lti:instrole:ims": cls.INSTITUTION,
+            "urn:lti:role:ims": cls.COURSE,
+            # LTI 1.3 https://www.imsglobal.org/spec/lti/v1p3/#lis-vocabulary-for-context-roles
+            "http://purl.imsglobal.org/vocab/lis/v2/system": cls.SYSTEM,
+            "http://purl.imsglobal.org/vocab/lis/v2/institution": cls.INSTITUTION,
+            "http://purl.imsglobal.org/vocab/lis/v2/membership": cls.COURSE,
+        }
+
+        for prefix, scope in scopes_prefixes.items():
+            if role.startswith(prefix):
+                return scope
+
+        # This is deprecated but allowed in LTI 1.3
+        # Conforming implementations MAY recognize the simple names for context roles;
+        # thus, for example, vendors can use the following roles interchangeably:
+        #   http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor
+        #   Instructor
+        # However, support for simple names in this manner for context roles is deprecated;
+        # by best practice, vendors should use the full URIs for all roles (context roles included).
+        #
+        # If we can't match any, pick the most restrictive
+        return cls.COURSE
```

Comment:
Ok. So we're able to get away with just using `startswith()` for our scope "parsing". We first look for LTI 1.1 scopes (`"urn:lti:..."), then 1.3 scopes `"http://purl.imsglobal.org/vocab/lis/v2/..."`), and if none of those are found then we return `"course"` because that's the most restrictive scope.

So this means if the role is completely unscoped (allowed in LTI 1.1, deprecated but allowed in 1.3) we'll return `"course"`. Also if there is a scope but we don't recognise it, we'll return `"course"`.

I think this is probably fine.

Is there a potential problem with a role coming with a scope that we don't recognise but this scope is *not* supposed to give the user access to the course at all, and we wrongly interpret the scope as `"course"` and grant them access?